web/swish/charts.ss

```diff
@@ -114,14 +116,23 @@
           (http:percent-encode chart)
           (http:percent-encode limit)) text))
 
-(let* ([chart (cond
-               [(find-param "c") => (lambda (x) x)]
-               [else (caar valid-charts)])]
-       [cols (cond
-              [(assoc chart valid-charts) => cdr]
-              [else (raise `#(invalid-chart ,chart))])]
-       [limit (or (find-param "limit") "-7 days")])
-  (with-db [db (log-file) SQLITE_OPEN_READONLY]
+(with-db [db (log-file) SQLITE_OPEN_READONLY]
+  (let* ([valid-charts
+          (append standard-charts
+            (map
+             (lambda (row)
+               (match row
+                 [#(,key)
+                  `(,key ,(format "json_extract(foreign_handles, '$.~a')" key))]))
+             (execute-sql db
+               "select distinct(key) from statistics, json_each(statistics.foreign_handles) order by key")))]
```

Comment:
Does this query walk through the entire statistics table? If so, please try it on a full statistics table based on the automatic pruning.

Reply:
If it's too slow, you could introduce a table to hold the foreign handle names and update the statistics table with an insert trigger that keeps this table up to date.

Reply:
Good catch!

I tried it with a full statistics table with one row every 5 minutes for 90 days. The original query was slow. When I changed it to grab the distinct keys via a subquery and moved the `order by` outside it sped up enough to be acceptable. The new query sped up even more when I changed it to operate on distinct `foreign_handles` so we skip redundant entries where we're at a steady state. I also tried it with an insert trigger along the way, but rolled it back once I found the `order by` mistake.

We'll eventually want to rewrite the diagnostics pages.

Reply:
Great!